pyamg/relaxation/smoothing.py

```diff
@@ -620,66 +676,150 @@ def smoother(A, x, b):
     return smoother
 
 
-def setup_gmres(lvl, tol=1e-12, maxiter=1, restrt=None, M=None, callback=None,
+def setup_cf_jacobi(lvl, f_iterations=DEFAULT_NITER, c_iterations=DEFAULT_NITER,
+                    iterations=DEFAULT_NITER, omega=1.0, withrho=False):
+    """Set up coarse-fine Jacobi."""
+    if withrho:
+        omega = omega/rho_D_inv_A(lvl.A)
+
+    Fpts, Cpts = _extract_splitting(lvl)
+
+    def smoother(A, x, b):
+        relaxation.cf_jacobi(A, x, b, Cpts=Cpts, Fpts=Fpts, f_iterations=f_iterations,
+                             c_iterations=c_iterations, iterations=DEFAULT_NITER,
+                             omega=omega)
+    return smoother
+
+
+def setup_fc_jacobi(lvl, f_iterations=DEFAULT_NITER, c_iterations=DEFAULT_NITER,
+                    iterations=DEFAULT_NITER, omega=1.0, withrho=False):
+    """Set up fine-coarse Jacobi."""
+    if withrho:
+        omega = omega/rho_D_inv_A(lvl.A)
+
+    Fpts, Cpts = _extract_splitting(lvl)
+
+    def smoother(A, x, b):
+        relaxation.fc_jacobi(A, x, b, Cpts=Cpts, Fpts=Fpts, f_iterations=f_iterations,
+                             c_iterations=c_iterations, iterations=DEFAULT_NITER,
+                             omega=omega)
+    return smoother
+
+
+def setup_cf_block_jacobi(lvl, f_iterations=DEFAULT_NITER, c_iterations=DEFAULT_NITER,
+                          iterations=DEFAULT_NITER, omega=1.0, Dinv=None, blocksize=None,
+                          withrho=False):
+    """Set up coarse-fine block Jacobi."""
+    # Determine Blocksize
+    if blocksize is None and Dinv is None:
+        if sparse.isspmatrix_csr(lvl.A):
+            blocksize = 1
+        elif sparse.isspmatrix_bsr(lvl.A):
+            blocksize = lvl.A.blocksize[0]
+    elif blocksize is None:
+        if sparse.isspmatrix_bsr(Dinv):
+            blocksize = Dinv.blocksize[1]
+        else:
+            blocksize = 1
+
+    # Check for compatible dimensions
+    if (lvl.A.shape[0] % blocksize) != 0:
+        raise ValueError('Blocksize does not divide size of matrix.')
+    if len(lvl.splitting)*blocksize != lvl.A.shape[0]:
+        raise ValueError('Blocksize not compatible with CF-splitting and matrix size.')
+
+    if blocksize == 1:
+        # Block Jacobi is equivalent to normal Jacobi
+        return setup_cf_jacobi(lvl, iterations=iterations, omega=omega, withrho=withrho)
+
+    Fpts, Cpts = _extract_splitting(lvl)
+
+    # Use Block Jacobi
+    if Dinv is None:
```

Comment:
I was thinking, wonder if we should store Dinv in the ml class or smoother class or something. Need to check more carefully, but I think if we call, e.g., CF block jacobi pre relax and FC block jacobi post relax (as we would do in AIR typically), we construct Dinv twice. It is easy for these blocks to get nontrivial in size, so might (especially in serial) be reasonable savings to avoid doing the dense inverses twice